ops/framework.py

```diff
@@ -328,19 +329,24 @@ def __str__(self):
 
 class SQLiteStorage:
 
+    DB_LOCK_TIMEOUT = timedelta(hours=1).total_seconds()
+
     def __init__(self, filename):
-        self._db = sqlite3.connect(str(filename), isolation_level="EXCLUSIVE")
+        # The isolation_level argument is set to None such that autocommit behavior of the sqlite3 module is disabled.
+        self._db = sqlite3.connect(str(filename), isolation_level=None, timeout=self.DB_LOCK_TIMEOUT)
         self._setup()
 
     def _setup(self):
-        c = self._db.execute("BEGIN")
+        # Make sure that the database is locked until the connection is closed, not until the transaction ends.
+        c = self._db.execute("PRAGMA locking_mode=EXCLUSIVE")
+        c = self._db.execute("BEGIN EXCLUSIVE")
```

Comment:
If I understand the SQLite documentation correctly, EXCLUSIVE shouldn't be required here if we have the pragma.

Reply:
https://github.com/canonical/operator/pull/141/commits/f6004043ef0519edfe7dee7aace2ccab22300d4a